src/main/daml/Daml/Finance/Interface/Util/SyntheticKey.daml

```diff
@@ -0,0 +1,254 @@
+  viewtype V
+
+-- | View for `I`.
+data View = View {}
+  deriving (Eq, Show)
+
+-- ============= State: Generic Index Table =============
+-- | Index table (used for looking up contract ids for keys).
+newtype IndexTable = IndexTable {getIndexTable : M.Map Text (ContractId SyntheticKey)} deriving (Eq,Show)
+
+-- | HIDE
+class
+  ( Eq k
+  , Show k
+  , Ord k
+  , HasFetch i
+  , HasInterfaceTypeRep i
+  , HasFromInterface i SyntheticKey
+  , HasToInterface i I
+  ) =>
+  HasSyntheticKey i k where
+    getKey : i -> k
+    getKeyFromCid : Party -> ContractId i -> Update k
+
+-- | Insert.
+insert : (HasSyntheticKey i k) => k -> ContractId i -> IndexTable -> IndexTable
+insert k cid idx =
+  IndexTable $ M.insert (sha256 $ show k) (toInterfaceContractId @I cid) (getIndexTable idx)
+
+-- | Delete.
+delete : (Show k) => k -> IndexTable -> IndexTable
+delete k idx = IndexTable $ M.delete (sha256 $ show k) (getIndexTable idx)
+
+-- | Lookup.
+lookup : (Show k) => k -> IndexTable -> Optional (ContractId I)
+lookup k idx = M.lookup @Text (sha256 $ show k) (getIndexTable idx)
+
+-- | Insert (with state).
+insertWithState : (HasSyntheticKey i k) => k -> ContractId i -> StateT IndexTable Update ()
+insertWithState k cid = put =<< insert k cid <$> get
+
+-- | Delete (with state).
+deleteWithState : (Show k) => k -> StateT IndexTable Update ()
+deleteWithState k = put =<< delete k <$> get
+
+-- | Lookup (with state).
+lookupWithState : (Show k) => k -> StateT IndexTable Update (Optional (ContractId I))
+lookupWithState k = lookup k <$> get
+
+-- HIDE
+instance Semigroup IndexTable where
+  (<>) a b = IndexTable $ getIndexTable a <> getIndexTable b
+
+-- HIDE
+instance Monoid IndexTable where
+  mempty = IndexTable M.empty
+  mconcat = foldl (<>) mempty
+
+-- ============= Utility Functions (for StateT IndexTable Udate) =============
+
+-- | Create (with state).
+createWithState :
+  forall i k c v.
+  ( HasSyntheticKey i k
+  , HasCreate c
+  , HasToInterface c i
+  ) => c -> StateT IndexTable Update (ContractId c)
+createWithState c = do
+  cid <- create c
+  insertWithState (getKey @i @k $ toInterface c) (toInterfaceContractId @i cid)
+  pure cid
+
+-- | Archive (with state).
+archiveWithState :
+  forall i k t v.
+  ( HasSyntheticKey i k
+  , HasArchive t
+  , HasFetch t
+  , HasInterfaceTypeRep t
+  , HasToInterface t i
+  ) =>
+  ContractId t -> StateT IndexTable Update ()
+archiveWithState cid = do
+  t <- fetch cid
+  let k = getKey @i @k (toInterface @i t)
+  cid <- fromSomeNote ("index table lookup for " <> show k <> " failed") . lookup k <$> get
+  archive $ coerceInterfaceContractId @t cid
+  deleteWithState k
+
+-- ============= Overlapping instances (for Update and StateT IndexTable Update) =============
+
+-- | HIDE
+class HasCreateM m t where
+  create : t -> m (ContractId t)
+
+-- | HIDE
+instance HasCreate t =>
+  HasCreateM Update t where
+    create = Prelude.create
+
+-- | HIDE
+instance {-# OVERLAPPABLE #-} HasCreate t =>
```

Comment:
Currently, `fetch` (for `StateT IndexTable Update`) and `fetch` (for `Update`) is currently not being used in the same file. For that purpose, it needs to be "overlapping" (c.f. `exercise` in the `Instruction`).

Reply:
I think it would compile anyway.

The only real overlapping instances we have is the following:
- instance {-# OVERLAPPING #-} HasExercise t c r => HasExerciseM UpdateWithKey t c r
- instance {-# OVERLAPPING #-} HasExercise t c (r, IndexTable) => HasExerciseM UpdateWithKey t (IndexTable -> c) r where

If you have the type `UpdateWithKey t (IndexTable -> c) r`, there are two matching instances for `exercise` and the compiler needs to be told which one it should prefer.

In the case of `fetch`, there is no overlapping instance because the types are different (`Update` vs `UpdateWithKey`). Even if these were used in the same file, the compiler should be able to disambiguate which one is the `fetch` that can be used.

This is to say that you probably need `OVERLAPPING` only on line 253 and can remove all the others

Reply:
that makes sense